controller/lib/src/etcd/mod.rs

```diff
@@ -0,0 +1,131 @@
+use etcd_client::{Client, DeleteResponse, Error, GetOptions, PutResponse};
+use log::{debug, error, info};
+
+pub struct EtcdClient {
+  inner: Client,
+}
+
+impl EtcdClient {
+  pub async fn new(address: String) -> Self {
+    info!("Starting ETCD client on {}", address);
+    EtcdClient {
+      inner: Client::connect([address], None).await.unwrap(),
+    }
+  }
+
+  pub async fn get(&mut self, key: &str) -> Result<String, Error> {
+    if let Some(kv) = self.inner.get(key, None).await?.kvs().first() {
+      let res = kv.value_str();
+      info!(
+        "Retrieving value in ETCD : Key \"{}\" is associated with value \"{}\"",
+        key,
+        res.as_ref().unwrap()
+      );
+      res.map(|s| s.to_string())
+    } else {
+      Err(Error::from(std::io::Error::new(
+        std::io::ErrorKind::NotFound,
+        "Key not found",
+      )))
+    }
+  }
+  pub async fn put(&mut self, key: &str, value: &str) -> Result<PutResponse, Error> {
+    info!(
+      "Inserting value in ETCD : Key \"{}\" associated with value \"{}\"",
+      key, value
+    );
+    self.inner.put(key, value, None).await
+  }
+  pub async fn patch(&mut self, key: &str, value: &str) -> Result<PutResponse, Error> {
+    match self.get(key).await {
+      Ok(_) => {
+        info!(
+          "Updating value in ETCD : Key \"{}\" associated with new value \"{}\"",
+          key, value
+        );
+        self.inner.put(key, value, None).await
+      }
+      Err(err) => Err(err),
+    }
+  }
+  pub async fn delete(&mut self, key: &str) -> Result<DeleteResponse, Error> {
+    if let Some(_) = self.inner.get(key, None).await?.kvs().first() {
+      info!("Deleting value in ETCD : Key \"{}\" ", key);
+      self.inner.delete(key, None).await
+    } else {
+      Err(Error::from(std::io::Error::new(
+        std::io::ErrorKind::NotFound,
+        "Key not found",
+      )))
+    }
+  }
```

Comment:
How is that different from simply calling `self.inner.delete(key, None).await` ?

Reply:
In this case, if the key is not found it doesn't throw an error.

Reply:
> In this case, if the key is not found it doesn't throw an error.

```rust
if let Some(_) = self.inner.get(key, None).await?
```

Wouldnt that return an error if the key is not found?

Reply:
Or do you mean that `self.inner.delete(key, None).await` would not return an error? Is that a problem?

Reply:
Yeah this command : `self.inner.delete(key, None).await` doesn't return an error.  
It could be debated but I was thinking it's more relevant to return one. Tell me if I'm wrong

Reply:
> Yeah this command : `self.inner.delete(key, None).await` doesn't return an error. It could be debated but I was thinking it's more relevant to return one. Tell me if I'm wrong

If you want to track the fact that a key already exists in the db, I'd suggest returning an `Option`. You're not supposed to know each and every key that's there, so asking for a kv to be deleted when/if k is not present is not an error.

Reply:
Okay, finally turning the function returning on `Option<DeleteResponse>`, should be the good one this time.

Thanks for you're time reviewing me